iron-input.html

```diff
@@ -12,300 +12,417 @@
-`prevent-invalid-input` and `allowed-pattern` attributes together to accomplish this. This feature
+`allowed-pattern` attribute to accomplish this. This feature
 is separate from validation, and `allowed-pattern` does not affect how the input is validated.
 
-    <!-- only allow characters that match [0-9] -->
-    <input is="iron-input" prevent-invalid-input allowed-pattern="[0-9]">
+    // Only allow typing digits, but a valid input has exactly 5 digits.
+    <iron-input allowed-pattern="[0-9]">
+      <input pattern="\d{5}">
+    </iron-input>
 
 @hero hero.svg
 @demo demo/index.html
-*/
+-->
+
+<dom-module id="iron-input">
+  <template>
+    <style>
+      :host {
+        display: inline-block;
+      }
+    </style>
+    <slot id="content"></slot>
+  </template>
+  <script>
+    Polymer({
+      is: 'iron-input',
 
-  Polymer({
+      behaviors: [
+        Polymer.IronValidatableBehavior
+      ],
 
-    is: 'iron-input',
+      /**
+       * Fired whenever `validate()` is called.
+       *
+       * @event iron-input-validate
+       */
 
-    extends: 'input',
+      properties: {
+
+        /**
+         * Use this property instead of `value` for two-way data binding, or to
+         * set a default value for the input. **Do not** use the distributed
+         * input's `value` property to set a default value.
+         */
+        bindValue: {
+          type: String
+        },
+
+        value: {
+          computed: '_computeValue(bindValue)'
+        },
+
+        /**
+         * Regex-like list of characters allowed as input; all characters not in the list
+         * will be rejected. The recommended format should be a list of allowed characters,
+         * for example, `[a-zA-Z0-9.+-!;:]`.
+         *
+         * This pattern represents the allowed characters for the field; as the user inputs text,
+         * each individual character will be checked against the pattern (rather than checking
+         * the entire value as a whole). If a character is not a match, it will be rejected.
+         *
+         * Pasted input will have each character checked individually; if any character
+         * doesn't match `allowedPattern`, the entire pasted string will be rejected.
+         *
+         * Note: if you were using `iron-input` in 1.0, you were also required to
+         * set `prevent-invalid-input`. This is no longer needed as of Polymer 2.0,
+         * and will be set automatically for you if an `allowedPattern` is provided.
+         *
+         */
+        allowedPattern: {
+          type: String
+        },
+
+        /**
+         * Set to true to auto-validate the input value as you type.
+         */
+        autoValidate: {
+          type: Boolean,
+          value: false
+        }
+      },
 
-    behaviors: [
-      Polymer.IronValidatableBehavior
-    ],
+      observers: [
+        '_bindValueChanged(bindValue, _inputElement)'
```

Comment:
Oh okay, I didn't know the properties block added extra work. I thought it optimized /shrug